src/cache/tlru-cache.ts

```diff
@@ -0,0 +1,56 @@
+import { LRUCache } from './lru-cache';
+
+/**
+ * Time-aware, least-recently-used (TLRU), variant of LRU where entries have valid lifetime.
+ * @param {number} maxSize - Maximum cache size
+ * @param {number} ttl - Time in milliseconds after which cache entry will evict itself
+ **/
+export class TLRUCache extends LRUCache {
+  private readonly cacheEntriesTimoutIds = new Map<string, ReturnType<typeof setTimeout>>();
+  constructor(readonly maxSize: number, readonly ttl: number) {
+    super(maxSize);
+  }
+
+  private clearCacheEntryTimeoutIfExists(key: string): void {
+    if (this.cacheEntriesTimoutIds.has(key)) {
+      const timeoutId = this.cacheEntriesTimoutIds.get(key);
+      clearTimeout(timeoutId);
+      this.cacheEntriesTimoutIds.delete(key);
+    }
+  }
+
+  private setCacheEntryTimeout(key: string): void {
+    const timeoutId = setTimeout(() => {
+      this.delete(key);
+    }, this.ttl);
```

Comment:
while conceptually this makes sense, I think in practice having a timeout at play for every cache entry is less efficient than simply storing when it expires and checking that when retrieving the entry. The timeouts will take up more system memory and could burdon the node event loop as well.

Instead, how about you set an expires (`e.g., Date.now() + expiresMs`) and then, on reads, you check and exclude expired entries in a "just in time" fashion.

Reads would be `get()`, `has()`, `entries()`, `values()`, and `keys()`. The latter are looking at the whole map so you could just do a full purge using a helper function before delegating to `super`.

On writes you'd just clobber the existing expires time.

Reply:
Superficially looked at [the TLRU paper](https://arxiv.org/pdf/1801.00390) and it looks like their suggestion is to perform eviction only on `set` operations. The gist of algorithm is basically:
- if there's space to insert an item, just insert it
- otherwise, evict entries that expired their TTL
- if there's no entries to evict by TTL, evict least-recently-used entry

This also the approach that python's [cachetools](https://github.com/tkem/cachetools/blob/6c78a8ff67b21834c3873cc918d87b2c5cb22e5c/src/cachetools/__init__.py#L376) uses — evict on sets and full-collection reads. For single-item reads, they do a simple filtering of result (`if now < entry ttl`) but don't remove the entry

Reply:
i've made changes to the implementation